Update Tips and Tricks documentation for raw / endraw tags

lhw-1 opened the pull request:
**What is the purpose of this pull request?**

- [X] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**

Updates the documentation for the `{% raw %}{% endraw %}` tags to be more reader friendly. This change is made in the [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) seciton.

This issue was initially brought up in #762, which was resolved by the introduction of the raw / endraw tags in #1220. As this was already resolved, the issue #1835 was modified from a bug report to a documentation update. This PR hence resolves #1835 by addressing the suggestion to make the documentation for this feature more reader-friendly.

<br>

**Anything you'd like to highlight/discuss:**

While editing this part of the documentation, I initially wanted to use the code / output boilerplate like the ones currently being used in the [Formatting Contents](https://markbind.org/userGuide/formattingContents.html) section, like such:

```html
<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code" id="list-example">
{{ '{{' }} '{{' {{ '}}' }} content {{ '{{' }} '}}' {{ '}}' }}
</variable>
</include>
```

However, this seems to result in a compilation error with the following error message:

```
Invalid expression: missing ) after argument list in
  _s(' {{ ')+" content "+_s(')+"  ' }}"
Raw expression: {{ ' {{ ' }} content {{ '  }}  ' }}
```

Meanwhile, this:

```html
<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code" id="list-example">
{{ '{% raw %}' }} content {{ '{% endraw %}' }}
</variable>
</include>
```

resulted in the following output on a test site:

![Screenshot from 2023-01-30 17-27-50](https://user-images.githubusercontent.com/25716529/215438802-6913a7eb-7e1b-4ccf-b086-e5d1711003d9.png)

As this is unrelated to the documentation update, it will be posted as a separate bug report. In any case, this is the reason why the code / output boilerplates were not used in this documentation update. (Any alternative suggestions would be great!)

Update: the issues #2119 and #2120 were opened to address these two bugs.

Apart from this, there are two more suggestions I would like input on, as they may / may not unnecessarily bloat the documentation:

1. Would it be helpful to include this section on [Support for Nunjucks](https://markbind.org/userGuide/markBindSyntaxOverview.html#support-for-nunjucks) as a dropdown / content fragment in the [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) section?

2. Would it be helpful to include this [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) section as a dropdown / content fragment in the [Math Formulae](https://markbind.org/userGuide/formattingContents.html#math-formulae) section, or any other relevant section?

<br>

**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [x] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

itsyme commented on docs/userGuide/tipsAndTricks.md:
```diff
@@ -34,16 +34,15 @@ For Markdown syntax: To display a literal character that are normally used for M
+This isn't necessary for `<code>` elements, markdown code fences and inline code though, which MarkBind automatically adds `v-pre` for.
 
-However, this does not change the need for `{% raw %}{% endraw %}`. Meaning, you can still use variables within your code!
+However, you will still need to use the `{% raw %} {% endraw %}` tags in order to use the variables within your code!
```
I think it would be better to avoid using personal pronouns like "you" when asking the user to do something, based on [Second Person](https://developers.google.com/style/person#:~:text=If%20you%27re%20telling,Click%20Submit.) in the Google developer documentation style guide. I think a good alternative could be "the `{% raw %} {% endraw %}` tags would still need to be used in order to use the variables within your code!".

lhw-1 replied:
Good suggestion! Fixed in [03eac18](https://github.com/MarkBind/markbind/pull/2118/commits/03eac1800464ebe163dfeff87aad3b4e57a026ce).

Thanks!

tlylt commented:
Some suggestions:
- reword this section to highlight how raw-endraw is used to escape nunjucks (in general), and highlight the more specific cases of `{{ content }}` and also for example the case in the original report `{% raw %}{% set title= "your title here" %}{% endraw %}`
- take this opportunity to explain how to add `v-pre` properly, instead of the currently unlinked `markdown-it-attrs` which can be confusing to readers who are unfamiliar to this package

> Would it be helpful to include this section on [Support for Nunjucks](https://markbind.org/userGuide/markBindSyntaxOverview.html#support-for-nunjucks) as a dropdown / content fragment in the [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) section?

👍 

> Would it be helpful to include this [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) section as a dropdown / content fragment in the [Math Formulae](https://markbind.org/userGuide/formattingContents.html#math-formulae) section, or any other relevant section?

There's already the following mention in the math formulae part, you can move this warning into the tips and then include the tips back in the math section
https://markbind.org/userGuide/formattingContents.html#math-formulae:~:text=If%20your%20equation%20requires%20special%20Nunjucks

lhw-1 commented:
Thanks for the detailed suggestions @tlylt - I agree with all of them. I'll update accordingly!

lhw-1 commented:
**Updates in commit [76ac122](https://github.com/MarkBind/markbind/pull/2118/commits/76ac122d5a1a0a8d00df02303e73775ca684c8d0)**:

> reword this section to highlight how raw-endraw is used to escape nunjucks (in general), and highlight the more specific cases of `{{ content }}` and also for example the case in the original report `{% raw %}{% set title= "your title here" %}{% endraw %}`

Explanations for why this syntax could sometimes be an issue was added (with reference to the built-in Nunjucks support in MarkBind). 

Which specific cases for `{{ content }}` did you have in mind @tlylt?

> take this opportunity to explain how to add `v-pre` properly, instead of the currently unlinked `markdown-it-attrs` which can be confusing to readers who are unfamiliar to this package

The phrase on `markdown-it-attr` was removed, and instead an example on how to add the `v-pre` as an HTML attribute has been added.

I've also included the section on [Support for Nunjucks](https://markbind.org/userGuide/markBindSyntaxOverview.html#support-for-nunjucks) as a link instead (as part of the explanation) in [Tips and Tricks - Using raw/endraw tags to display contents](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content) section, and modified the [Math Formulae](https://markbind.org/userGuide/formattingContents.html#math-formulae) section according to the suggestion by @tlylt above.

In addition, I think that it may be useful to include this also as a tip box within the [Formatting Contents - Code](https://markbind.org/userGuide/formattingContents.html#code) section, as this syntax was an issue I encountered while converting the Vue.js documentation to MarkBind, and it was mostly the code blocks that had the `{{ content }}` syntax. Any suggestions on where within this section to put this tip box?

tlylt commented:
> > reword this section to highlight how raw-endraw is used to escape nunjucks (in general), and highlight the more specific cases of `{{ content }}` and also for example the case in the original report `{% raw %}{% set title= "your title here" %}{% endraw %}`
> 
> Explanations for why this syntax could sometimes be an issue was added (with reference to the built-in Nunjucks support in MarkBind).
> 
> Which specific cases for `{{ content }}` did you have in mind @tlylt?

I'm referring to exactly this one case of `{{ content }}`, as in this use of curly brackets for variables.

> In addition, I think that it may be useful to include this also as a tip box within the [Formatting Contents - Code](https://markbind.org/userGuide/formattingContents.html#code) section, as this syntax was an issue I encountered while converting the Vue.js documentation to MarkBind, and it was mostly the code blocks that had the `{{ content }}` syntax. Any suggestions on where within this section to put this tip box?

Should be ok at the end of the section.

lhw-1 commented:
**Updates in commit [22379a3](https://github.com/MarkBind/markbind/pull/2118/commits/22379a35a0d118cfaa2016a11b6744d3867c1b67)**:

I've added a more specific example of `{{ username }}`, taken from the Nunjucks documentation. I've also added in a new subsection under the [Formatting Contents - Code](https://markbind.org/userGuide/formattingContents.html#code) section on displaying `{{ content }}` raw.

tlylt commented on docs/userGuide/tipsAndTricks.md:
```diff
@@ -34,18 +34,33 @@ For Markdown syntax: To display a literal character that are normally used for M
 
 ##### :fas-lightbulb: Using {% raw %}{% endraw %} to display `{{ content }}`
 
+By default, MarkBind processes any code in the form of `{{ content }}`. This is because Nunjucks (which is [supported by MarkBind](https://markbind.org/userGuide/markBindSyntaxOverview.html#support-for-nunjucks)) uses this syntax to [evaluate an expression, variable, or function calls](https://mozilla.github.io/nunjucks/templating.html#variables). For instance:
```
Let's fix this link to use relative link so that it does not link out of the PR preview to markbind.org

lhw-1 replied:
Fixed in 5789e6e!

tlylt commented on docs/userGuide/tipsAndTricks.md:
````diff
@@ -34,18 +34,33 @@ For Markdown syntax: To display a literal character that are normally used for M
+{% endraw %}
+```
+
+Otherwise, the content will not be displayed. 
````
What does "the content" mean here?

lhw-1 replied:
Ah, the "content" refers to the HTML content that are enclosed in double curly braces. I've fixed the wording in [f5a6772](https://github.com/MarkBind/markbind/pull/2118/commits/f5a6772a019b07a60cc8ea4a83033d3fcac37f5a)!

lhw-1 commented on docs/userGuide/syntax/code.md:
```diff
@@ -228,6 +228,15 @@ or the java code `public static void main(String[] args)`{.java}.
 </variable>
 </include>
 
+##### Displaying `{{ content }}`
```
Ah, it's not a bug, I ironically forgot to include the raw / endraw tags :sweat_smile: 

As for `{{ content }}`, it does refer to the content within double curly brackets - I will change the title to something like "Displaying content within curly braces `{{ }}`" to make it much clearer!

lhw-1 replied:
Fixed in 5789e6e!

lhw-1 commented on docs/userGuide/syntax/code.md:
```diff
@@ -228,6 +228,15 @@ or the java code `public static void main(String[] args)`{.java}.
 
+##### Displaying `{{ content }}`
+
+If your equation requires special Nunjucks tags like {% raw %}`{{`{% endraw %} or {% raw %}`}}`{% endraw %},use a
```
Fixed in 5789e6e!

lhw-1 commented on docs/userGuide/syntax/code.md:
```diff
@@ -228,6 +228,15 @@ or the java code `public static void main(String[] args)`{.java}.
+##### Displaying `{{ content }}`
+
+If your equation requires special Nunjucks tags like {% raw %}`{{`{% endraw %} or {% raw %}`}}`{% endraw %},use a
+[raw-endraw block](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content):
```
Fixed in 5789e6e!

lhw-1 commented on docs/userGuide/syntax/mathformulae.md:
```diff
@@ -33,12 +33,9 @@ Finally, the Pythagoras theorem: $c^2 = a^2 + b^2$.
+<box type="info">
 
 If your equation requires special Nunjucks tags like {% raw %}`{{`{% endraw %} or {% raw %}`}}`{% endraw %},use a
 [raw-endraw block](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content):
```
Fixed in 5789e6e!

lhw-1 commented on docs/userGuide/tipsAndTricks.md:
```diff
@@ -34,18 +34,33 @@ For Markdown syntax: To display a literal character that are normally used for M
 
-This isn't necessary for `<code>` elements, markdown code fences and inline code though, which MarkBind automatically
-adds `v-pre` for.
+If using raw-endraw tags on html content, you would also need to add the `v-pre` attribute as an HTML attribute, like this:
```
Fixed in 5789e6e!

tlylt commented on docs/userGuide/syntax/code.md:
```diff
@@ -228,6 +228,17 @@ or the java code `public static void main(String[] args)`{.java}.
+
+##### Displaying content within curly braces: `{{ content }}`
+
+If your code contains special Nunjucks tags like `{{` or `}}`,use a [raw-endraw block](../tipsAndTricks.html#using-raw-endraw-to-display-content):
```
```suggestion
If your code contains special Nunjucks tags like `{{` or `}}`, use a [raw-endraw block](../tipsAndTricks.html#using-raw-endraw-to-display-content):
```

lhw-1 replied:
Fixed in [0b0701a](https://github.com/MarkBind/markbind/pull/2118/commits/0b0701af78d7a1071712a97fa26c0c3803d50867), along with the same error in `mathformulae.md` in the same directory!

tlylt commented on docs/userGuide/tipsAndTricks.md:
````diff
@@ -34,18 +34,33 @@ For Markdown syntax: To display a literal character that are normally used for M
+```
+{% raw %}
+
+<p v-pre>{{ content }}</p>
````
Can we change this to `<div></div>` instead, to avoid hydration errors. Also, what about the markdown-it-attrs method previously mentioned? You removed it because it won't work or?

lhw-1 replied:
Can I confirm that when the `v-pre` attribute is added, compilation is skipped on the Vue side for the element and its children as per the docs [here](https://vuejs.org/api/built-in-directives.html#v-pre), so even if a `<p>` tag is used, hydration errors should not occur since `v-pre` attribute has been added? 

This is just a side question, I'll add in a `<div></div>` to follow conventions :+1: 

Also, as for the `markdown-it-attrs` method, I initially removed it as I thought it was not working with the current implementation, but after some digging and testing, it still seems to work - I will be adding in a link and an example for this as well!

tlylt reviewed (changes requested):
Thanks @lhw-1 for applying the changes, some more nits.

tlylt commented on docs/userGuide/tipsAndTricks.md:
````diff
@@ -34,18 +34,47 @@ For Markdown syntax: To display a literal character that are normally used for M
+{% endraw %}
+```
 
+For Markdown elements, the `v-pre` attribute needs to be declared using [`markdown-it-attrs`](syntax/attributes.html#classes-attributes-and-amp-identifiers), which allows classes, identifiers, and attributed to be added to Markdown syntax:
````
https://deploy-preview-2118--markbind-master.netlify.app/userGuide/syntax/attributes.html#classes-attributes-and-amp-identifiers

This link 404.

lhw-1 replied:
Fixed in [1ada4f3](https://github.com/MarkBind/markbind/pull/2118/commits/1ada4f384a5d865f2423b970f58c8334ccc7a003)!

tlylt commented:
@lhw-1 pls remember to include the Proposed commit message in the future.
